### PR TITLE
feat(use cases): Add "Good Error Messages"

### LINKED ISSUE #6: Document use case: Good error messages for mistaken definitions

### PR DESCRIPTION
**Pull request by DavHau**
This PR enhances the use case documentation for "Good Error Messages", providing a concrete example and outlining the current issues. This serves as a guidepost for future improvement initiatives in error messaging within the Nix ecosystem.

closes https://github.com/nixpkgs-architecture/pkgs-modules/issues/6

### CODE REVIEW

**Comment by infinisil** [comment]
I'm thinking this is a goal rather than a use case, and thinking about how the two relate to each other, I think that our goal should be to fix the problems of the use cases.

How about moving this to a `problems` directory, then linking this item to that file: https://github.com/nixpkgs-architecture/pkgs-modules/blob/ac18343757b5182d2d749748723215261e5ff5c0/use-cases/Customizing%20a%20package%20definition.md#L36

**Review by infinisil — CHANGES_REQUESTED** [review]
(Test to see what happens when a team is requested to review a PR and a review from a member is made). Edit: Turns out it doesn't request reviews from the rest of the team then. This seems to be a popular feature request: https://github.com/orgs/community/discussions/5289

**Review by alex-ameen — APPROVED** [review]

**Comment by infinisil** [comment]
Oh I see, #18 was a PR against this branch :smile:

**Review by phaer — APPROVED** [review]
I think @infinisil proposal to move it to a `problems` directory is good, as is moving it to `goals`. 

Content LGTM :)

**Review by infinisil — APPROVED** [review]

**Comment by DavHau** [comment]
Moved the file and added a link from the customization document.

File: problems/good-error-messages.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+## Current Problems
+
+1. Lack of type safety: Currently, users can provide incorrect types when defining attributes, causing confusion without any explicit error messages. For instance, a user might try to set `dontUnpack = "false"`, expecting it to behave the same as `dontUnpack = false`. However, due to the lack of type safety, the package still builds, albeit with unexpected behavior.
+2. Absence of name checking: In the current ecosystem, if a user mistakenly types an incorrect attribute name, there is no mechanism to check and alert the user of this mistake. This can lead to situations where a misspelled attribute doesn't have the desired effect, and no errors are produced to guide the user. For instance, if a user types buildPhases instead of buildPhase, the intended build phases will not be executed, yet the system will not report an error, causing confusion and potential misbehavior in the package.
+3. Unknown code origin: For build time errors it is often unclear where the code defining the failing logic is located. For example it is unclear if a failing command originates from the users overrides or from a setup hook that is included elsewhere.
```
Comment (edolstra) [inline problems/good-error-messages.md:18]: Another issue (or maybe an expansion of issue 3): Error messages should be given as early as possible. Currently type errors often manifest at a late stage (e.g. deep in a recursive use of a `lib` function), where they could have been presented much earlier and with a better error message if package function arguments were typed.

**Review by edolstra — APPROVED** [review]

File: problems/good-error-messages.md (hunk 1 above)
Comment (DavHau) [reply problems/good-error-messages.md:18]: added as an additional problem via 7c33270946b9ea7b7be27cb6108d4e51d17e1a2d

**Comment by DavHau** [comment]
Via https://github.com/nixpkgs-architecture/pkgs-modules/pull/16/commits/d5ee0cdfc023c261724912d21d200ccce79cbef0 I removed the following statement from the `General Information` section:

`For instance, if a package fails to build due to a missing dependency, the error message should clearly identify the missing dependency and where it's expected.`

... as this is more like a build time logic problem and I'm not sure if this is in scope for us.

**Pull request merged**